src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509CertificateReader.cs

```diff
@@ -0,0 +1,416 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal sealed class OpenSslX509CertificateReader : ICertificatePal
+    {
+        private static DateTimeFormatInfo s_validityDateTimeFormatInfo;
+
+        private SafeX509Handle _cert;
+        private X500DistinguishedName _subjectName;
+        private X500DistinguishedName _issuerName;
+
+        internal unsafe OpenSslX509CertificateReader(byte[] data)
+        {
+            fixed (byte* pDataFixed = data)
+            {
+                byte* pData = pDataFixed;
+                byte** ppData = &pData;
+
+                _cert = Interop.libcrypto.d2i_X509(IntPtr.Zero, ppData, data.Length);
+
+                // X509_check_purpose has the effect of populating the sha1_hash value,
+                // and other "initialize" type things.
+                bool init = Interop.libcrypto.X509_check_purpose(_cert, -1, 0);
+
+                if (!init)
+                {
+                    throw new CryptographicException(Interop.libcrypto.GetOpenSslErrorString());
+                }
+            }
+        }
+
+        public bool HasPrivateKey
+        {
+            get { return false; }
+        }
+
+        public AsymmetricAlgorithm PrivateKey
+        {
+            get { return null; }
+        }
+
+        public IntPtr Handle
+        {
+            get { return IntPtr.Zero; }
+        }
+
+        public string Issuer
+        {
+            get { return IssuerName.Name; }
+        }
+
+        public string Subject
+        {
+            get { return SubjectName.Name; }
+        }
+
+        public byte[] Thumbprint
+        {
+            get
+            {
+                int negativeSize = Interop.NativeCrypto.GetX509Thumbprint(_cert, null, 0);
```

Comment:
This pattern looks very much like the GetAsn1StringBytes where the actual P/Invoke is private and a wrapper is provided in the interop type.  Should the same be done for this in order to be consistent?